user_guide/docs/settings.md
To use this editor:

1. Connect your Castor & Pollux to your Eurorack power supply. The module **must** be powered by the Eurorack power supply.
1. Connect a USB cable from your computer to Castor & Pollux's USB port. It's located on the right side of the module on the bottom circuit board.
1. Click the ++"connect"++ button below, and when prompted, allow the page access to your MIDI devices.
1. Once connected the editor will appear and load the current settings from your device.
1. Once you're done editing, click the ++"save & restart"++ button at the bottom of the editor.
1. Saving should automatically reset your module, but, if you don't see the changes take effect then try restarting your module. You can do this by turning it off and back on or by pressing the small reset button that's right beside the USB port.

!!! Note
    This editor uses [WebMIDI](https://www.midi.org/17-the-mma/99-web-midi) to communicate with Castor & Pollux. It works in [Google Chrome](https://www.google.com/chrome/), [Microsoft Edge](https://www.microsoft.com/en-us/edge), and [Opera](https://www.opera.com/) ([full list of supported browsers](https://caniuse.com/midi)).

If you run into issues, feel free to [reach out](mailto:support@winterbloom.com).


<button class="button is-primary is-large" type="button" id="connect">Connect</button>

<div id="connect_info" class="admonition danger" hidden>
  <p class="admonition-title">Connection info</p>
</div>

<div id="firmware_incompatible" class="admonition fairy" hidden>
  <p class="admonition-title">Firmware update required</p>
  <p>You'll need to update your module's firmware to use this editor. Don't worry, it's really easy! <a href="https://github.com/wntrblm/Castor_and_Pollux/releases/latest" target="_blank">Get the newest firmware</a>
</div>

<form class="form" id="settings_editor">
  <!-- Device information -->
  <section id="info_section" hidden>
    <h2>Device info</h2>
    <div class="grid">
      <div>
        <label for="firmware_version">Firmware version</label>
        <div role="group">
          <input id="firmware_version" type="text" name="firmware_version" readonly value="" />
          <button type="button" title="copy to clipboard" data-clipboard-copy-target="firmware_version">
              <winter-icon>content_copy</winter-icon></button>
        </div>
        <label for="hardware_revision">Hardware revision</label>
        <input id="hardware_revision" type="text" name="hardware_revision" readonly value="" />
        <label for="serial_number">Serial number</label>
        <div role="group">
          <input id="serial_number" type="text" name="serial_number" readonly value="" />
          <button type="button" title="copy to clipboard" data-clipboard-copy-target="serial_number">
            <winter-icon>content_copy</winter-icon></button>
        </div>
      </div>
      <aside>
        This is your C&P's current firmware version, hardware revision, and serial number. We might ask you for this information if you reach out for help. C&PI is hardware revisions <code>1-4</code> and C&PII is hardware revisions <code>5</code> and above.
      </aside>
    </div>
  </section> <!-- info_section -->
  <!-- General settings -->
  <section id="settings_section" hidden>
    <h2>Settings</h2>
    <div class="grid col-2">
      <div>
        <label for="led_brightness">LED brightness</label>
        <input type="range" name="led_brightness" min="0" max="254" value="127" data-bind data-bind-type="int" />
        <small><span data-display-value-for="led_brightness" data-display-formatter="(input.valueAsNumber / 254 * 100).toFixed(0)"></span> percent</small>
      </div>
      <aside>
        The higher the brightness, the more power the module will consume. At maximum brightness, it consumes <code>130mA</code> from the <code>+12V</code> rail.</aside>
      <div>
        <label for="base_cv_offset">Base pitch CV offset</label>
        <input type="number" name="base_cv_offset" value="1.0" step="0.083" min="0" max="5" data-bind data-bind-type="float" data-binding-precision="2" />
        <small>octaves (volts)</small>
      </div>
      <aside>
        This setting effectively determines the starting note for the module since the pitch <abbr title="Control voltage">CV</abbr> input and offset knobs are added to this value to determine the final pitch CV. By default, it's <code>1.0</code>.
      </aside>
      <div>
        <label for="castor_knob_min">Castor pitch knob minimum</label>
        <input type="number" name="castor_knob_min" value="-1.02" step="0.1" min="-3.3" max="0" data-bind data-bind-type="float" />
        <small>octaves (volts)</small>
        <label for="castor_knob_max">Castor pitch knob maximum</label>
        <input type="number" name="castor_knob_max" value="1.10" step="0.1" min="0" max="3.3" data-bind data-bind-type="float" />
        <small>octaves (volts)</small>
        <label for="pollux_knob_min">Pollux pitch knob minimum</label>
        <input type="number" name="pollux_knob_min" value="-1.02" step="0.1" min="-3.3" max="0" data-bind data-bind-type="float" />
        <small>octaves (volts)</small>
        <label for="pollux_knob_max">Pollux pitch knob maximum</label>
        <input type="number" name="pollux_knob_max" value="1.10" step="0.1" min="0" max="3.0" data-bind data-bind-type="float" />
        <small>octaves (volts)</small>
      </div>
      <aside>
        These settings determine the pitch offset knob's range for each oscillator. By default, they have a range of about <code>-1</code> to <code>+1</code> octaves (volts). The value of the pitch offset knob is added to the pitch <abbr title="Control voltage">CV</abbr> input. Note that because the module's <abbr title="Analog to digital converter">ADC</abbr> isn't perfect, the knobs don't quite reach 100% of their range. You may need to nudge these to give the knob some headroom.
      </aside>
      <div>
        <label for="pitch_knob_nonlinearity">Pitch knob non-linearity</label>
        <input type="range" name="pitch_knob_nonlinearity" value="0.6" step="0.01" min="0.33" max="1.0" data-bind data-bind-type="float" />
        <small>
          <span data-display-value-for="pitch_knob_nonlinearity" data-display-format="percent"></span> percent</small>
      </div>
      <aside>
          The firmware adjusts the pitch <abbr title="Control voltage">CV</abbr> knobs so that it's easier to tune the oscillators. It does this by spreading the values near the center of the knob out so that the range at the center is more fine. This makes the knobs non-linear. This setting controls how "intense" that non-linearity is. <code>33%</code> is equivalent to a linear response. Higher values make it easier and easier to tune the oscillator but sacrifices the range, values lower than <code>33%</code> will make it harder to tune and aren't recommended. This is <code>60%</code> by default.
      </aside>
      <div>
        <label>Quantization</label>
        <label for="quantization_enabled">
          <input type="checkbox" id="quantization_enabled" name="quantization_enabled" value="on" data-bind />
          Enable Castor's semitone quantization
        </label>
      </div>
      <aside>
        When nothing is patched into Castor's pitch <abbr title="Control voltage">CV</abbr> input, Castor's pitch knob becomes a <em>coarse</em> pitch knob that sweeps across <code>6</code> octaves. By default, in this situation Castor will quantize to the nearest semitone. You can disable this to get continuous frequency changes.
      </aside>
      <div>
        <label>Jack detection</label>
        <label for="zero_detection_enabled">
          <input type="checkbox" id="zero_detection_enabled" name="zero_detection_enabled" value="on" data-bind />
          Enable pitch jack detection
        </label>
        <label for="zero_detection_threshold">Zero detection threshold</label>
        <input type="number" name="zero_detection_threshold" min="0" max="800" value="100" data-bind data-bind-type="int" />
        <small><span data-display-value-for="zero_detection_threshold"></span> code points
            </small>
      </div>
      <aside>
        These settings determine how C&P detects the presence or absence of a signal at the pitch <abbr title="Control voltage">CV</abbr> jacks. C&P does this by checking if the signal at the jack is <code>0V</code>.
        <br/><br/>
        The intended behavior is that when no cable is patched to Castor's pitch input, it'll switch to "coarse" pitch behavior. Likewise for Pollux, if there's no cable patched to its pitch input then it will follow Castor. However, some users may find this behavior undesirable, especially if you often send pitch CV of around <code>0V</code>.
      </aside>
      <div>
        <label for="chorus_max_intensity">Chorus maximum intensity</label>
        <input type="range" name="chorus_max_intensity" value="0.05" step="0.01" min="0" max="1.0" data-bind data-bind-type="float" />
        <small><span data-display-value-for="chorus_max_intensity" data-display-format="percent"></span> percent</small>
      </div>
      <aside>
        This setting controls the <strong>maximum</strong> intensity of the chorusing feature. The intensity is the amount that the internal <abbr title="Low-frequency oscillator">LFO</abbr> affects Pollux's pitch.
      </aside>
      <div class="column field">
        <label for="lfo_1_frequency">LFO frequency</label>
        <input type="range" name="lfo_1_frequency" value="0.2" step="0.1" min="0.1" max="10.0" data-bind data-bind-type="float"/>
        <small><span data-display-value-for="lfo_1_frequency"></span> hertz</small>
      </div>
      <aside>
          This setting controls the <strong>default</strong> frequency of the internal <abbr title="Low-frequency oscillator">LFO</abbr>. This LFO is used by the chorusing feature and the <abbr title="Pulse width modulation">PWM</abbr> feature. The frequency can be changed anytime by holding the <em>hard sync</em> button and moving the <em>chorusing amount</em> (<code>φ</code>) knob, this just sets the frequency at start-up.
      </aside>
      <div>
        <label class="label">LFO waveform</label>
        <canvas id="lfo-waveform-canvas" width="640" height="200" class="waveform teal"></canvas>
        <div class="grid col-2">
          <div>
            <label for="lfo_1_waveshape">Shape 1</label>
            <select id="lfo_1_waveshape" name="lfo_1_waveshape" data-bind data-bind-type="int">
              <option value="0">Triangle</option>
              <option value="1">Sine</option>
              <option value="2">Sawtooth</option>
              <option value="3">Square</option>
            </select>
            <label for="lfo_1_factor">Amplitude</label>
            <input type="range" id="lfo_1_factor" name="lfo_1_factor" value="1.0" step="0.01" min="0.0" max="1.0" data-bind data-bind-type="float"/>
          </div>
          <div>
            <label for="lfo_2_waveshape">Shape 2</label>
            <select id="lfo_2_waveshape" name="lfo_2_waveshape" data-bind data-bind-type="int">
              <option value="0">Triangle</option>
              <option value="1">Sine</option>
              <option value="2">Sawtooth</option>
              <option value="3">Square</option>
            </select>
            <label for="lfo_2_factor">Amplitude</label>
            <input type="range" id="lfo_2_factor" name="lfo_2_factor" value="1.0" step="0.01" min="0.0" max="1.0" data-bind data-bind-type="float"/>
            <label for="lfo_2_frequency_ratio">Frequency ratio</label>
            <input type="range" id="lfo_2_frequency_ratio" name="lfo_2_frequency_ratio" value="1.0" step="0.01" min="0.0" max="10.0" data-bind data-bind-type="float"/>
          </div>
        </div>
      </div>
      <aside>
          These settings control the overall waveshape of the internal <abbr title="Low-frequency oscillator">LFO</abbr>. The LFO is generated by adding two separate waveforms together. By default, the first waveform is a triangle wave and the second waveform is turned off.
      </aside>
      <div>
        <label for="pulse_width_bitmask">Pulse width behavior</label>
        <select id="pulse_width_bitmask" name="pulse_width_bitmask" data-bind data-bind-type="int">
            <option value="4095">Smooth</option>
            <option value="3840">Classic</option>
        </select>
      </div>
      <aside>
          This controls the behavior of the pulse wave when the pulse width is modulated. Smooth is almost certainly what you want, however, if you choose Classic you'll introduce some interesting glitchy/stepping behavior when the pulse width changes.
          <br>
          The reason this setting exists is because prior to the July 2022 firmware release, C&P had a bug that caused the effective pulse width resolution to be far lower than intended. Some folks like the old sound, so we added this setting for nostalgia.
      </aside>
    </div>
    <button type="submit" id="save_button">Save &amp; restart</button>
  </section><!-- settings_section -->
  <!-- Dangerous settings -->
  <section id="danger_section" hidden>
    <h2>Danger zone</h2>
    <p>These settings are calibrated during assembly and <strong>do not have default values</strong>. Changing these settings can cause your module to behave incorrectly, so do not touch them unless you know what you're doing or have been asked to do so by support. It's highly recommended that you write down these settings before making changes in case you need to restore them.</p>
    <label for="allow_danger">
      <input class="checkbox" type="checkbox" id="allow_danger" />
      I understand, let me in!
    </label>
    <div id="danger_section_content" class="grid col-2" hidden>
      <div>
        <label for="osc8m_freq">8 MHz clock measured frequency</label>
        <input type="number" name="osc8m_freq" min="7000000" max="9000000" value="8000000" data-bind data-bind-type="int"/>
        <small>Hertz</small>
      </div>
      <aside>
          C&P uses an internal <code>8 MHz</code> clock to generate the frequency signals for the oscillators. The actual frequency of this oscillator can vary between <code>7.94 MHz</code> and <code>8.06 MHz</code>. This setting is set to compensate for that so that the oscillator outputs are in tune.
      </aside>
      <div>
        <label for="cv_gain_error">CV input gain correction</label>
        <input type="number" name="cv_gain_error" value="1" min="0.5" max="2.0" step="0.0001" data-bind data-bind-type="float" data-binding-precision="4"/>
        <small>multiplier</small>
        <label for="cv_offset_error">CV input error correction</label>
        <input type="number" name="cv_offset_error" value="0" data-bind-type="float" data-bind data-binding-precision="1"/>
        <small>code points</small>
        <label for="adc_gain_corr">ADC gain correction</label>
        <input type="number" name="adc_gain_corr" value="2048" min="1024" max="3072" data-bind data-bind-type="int"/>
        <small><span data-display-value-for="adc_gain_corr" data-display-formatter="(input.valueAsNumber / 2048).toFixed(3)"></span>x multiplier</small>
        <label for="adc_offset_corr">ADC offset correction</label>
        <input type="number" name="adc_offset_corr" value="0" min="-100" max="100" data-bind data-bind-type="int" />
        <small>code points</small>
      </div>
      <aside>
        <p>
          These settings determine how the module corrects for gain and offset errors in its <abbr title="Analog to digital converter">ADC</abbr>. We calibrate every unit during manufacturing and save the calibration data to the device. However, we keep a copy of it in case it somehow gets erased or corrupted. You can use the restore button below to load the values that were originally on your device. If the button is greyed out, then we don't have a copy of your module's calibration data. ☹️
        </p>
        <button class="outline" type="button" id="restore_adc_calibration">Restore to factory values</button>
      </aside>
      <div>
        <label>Tuning calibration</label>
        <div class="grid col-2">
          <button type="button" class="secondary" id="measure_one">Measure 1V</button>
          <input id="measure_one_result" type="number" placeholder="not measured" readonly>
          <button type="button" class="secondary" id="measure_three">Measure 3V</button>
          <input id="measure_three_result" type="number" placeholder="not measured" readonly>
          <div>
            <label for="gain_error_result">Gain error</label>
            <input id="gain_error_result" name="gain_error_result" type="number" readonly/>
          </div>
          <div>
            <label for="offset_error_result">Offset error</label>
            <input id="offset_error_result" name="offset_error_result" type="number" readonly/>
          </div>
        </div>
        <button type="button" class="danger outline" id="save_tuning" disabled>Save</button>
      </div>
      <aside>
        This allows you to re-calibrate your module's tuning. You should only do this if you absolutely know what you're doing or have been instructed to do so by support. <strong>Please write down the values for the settings in the previous section</strong> as this will overwrite them.
        <br>
        To use this, connect the most accurate <abbr title="Control voltage">CV</abbr> source that you have to the first oscillator's pitch input. Set the CV to <code>1v</code> and click <kbd>Measure 1V</kbd>. Repeat for <code>3v</code>. Once both measurements have been taken, it will calculate new values for tuning calibration. Click <kbd>Save</kbd> to write these to the device.
      </aside>
    </div>
  </section><!-- danger_section -->
  <section id="monitor_section" hidden>
    <h2>Monitor</h2>
    <p>This is an advanced debugging tool. You shouldn't even see this unless you know what you're doing!</p>
    <button class="outline danger" type="button" id="enable_monitor">Enable monitoring</button>
    <pre class="code" id="monitor_output">monitor output</pre>
  </section>
  <section id="ramp_section" hidden>
    <h2>Ramp calibration</h2>
    <button class="button is-warning" type="button" id="ramp_swap">Swap ramp calibration</button>
    <pre class="code" id="ramp_output">Output will appear here</pre>
  </section>
</form>

<script type="module" src="../scripts/settings.js"></script>


hardware/CHANGELOG.md
# June 13th, 2021 (mainboard v4)

This hardware revision was largely just to add additional information to the silk screen and update MPNs. This revision is used in production round three.

* Added CE and WEEE logos to the mainboard.
* Assigned new MPNs on the mainboard.
* Added a laser-cuttable jig for the programming board.
* Added missing values to jackboard potentiometers.
* Pulled the copper fills on the mainboard back from the board edge.
* Pulled the copper fills on the programming board back from the board edge.
* Updated hierarchal symbols directions with proper signal directions.
* Created new version of the programming board with significantly more test points.

# January 27, 2021 (mainboard v3)

This hardware revision was created to obtain better performance out of the ramp oscillators.

* Changed license from CC BY SA to CERN-OHL-P v2.
* Changed the oscillator behavior: Q1 was changed from 2N3094 (NPN) to 3N3906 (PNP). The previous design unintentionally used Q1 in reverse-active mode which had a significant impact on the maximum ramp voltage that could be achieved. Changing this to an appropriate PNP allows the transistor to always work in either cutoff or forward-active mode. This allows the ramp voltage to climb as high as the output rail on U3B (11.3v). This greatly improves headroom and lowers the amount of noise audible from the ramp oscillator.
* Modified U3C to be an inverting attenuator instead of an inverting amplifier. This is because previously the ramp's output was limited to about -500mV, but now the ramp's target voltage is 10v. This attenuator drops it down to +3.3v so that the ramp comparators continues to work as expected.
* Added R60 & R61 to limit the amount of current when C9/C28 is discharging through Q1/Q2.
* Changed the values of C6/C25, C9/C28, R2/R26, R3/R27, R4/R28 to match the Juno 106's values.
* Added ESD warning symbol.

# September 26th, 2020

* Initial hardware version


scripts/settings.js
/*
    Copyright (c) 2021 Alethea Katherine Flowers.
    Published under the standard MIT License.
    Full text available at: https://opensource.org/licenses/MIT
*/

import { MIDI } from "/winter.js";
import { $e, $on } from "./utils.js";
import * as forms from "./forms.js";
import GemSettings from "./gem_settings.js";
import Gemini from "./gemini.js";
import GitHub from "./github.js";

const ui = {
    settings_form_elem: $e("settings_editor"),
    info_section: $e("info_section"),
    settings_section: $e("settings_section"),
    save_btn: $e("save_button"),
    allow_danger: $e("allow_danger"),
    danger_section: $e("danger_section"),
    dangerous_section_content: $e("danger_section_content"),
    connect_btn: $e("connect"),
    connect_info: $e("connect_info"),
    firmware_version: $e("firmware_version"),
    firmware_outdated: $e("firmware_outdated"),
    firmware_incompatible: $e("firmware_incompatible"),
    serial_number: $e("serial_number"),
    hardware_revision: $e("hardware_revision"),
    restore_adc_calibration_btn: $e("restore_adc_calibration"),
    lfo_waveform_canvas: $e("lfo-waveform-canvas"),
    tuning: {
        measure_one: $e("measure_one"),
        measure_one_result: $e("measure_one_result"),
        measure_three: $e("measure_three"),
        measure_three_result: $e("measure_three_result"),
        gain_error_result: $e("gain_error_result"),
        offset_error_result: $e("offset_error_result"),
        save: $e("save_tuning"),
    },
    monitoring: {
        section: $e("monitor_section"),
        enable: $e("enable_monitor"),
        output: $e("monitor_output"),
    },
    ramp: {
        section: $e("ramp_section"),
        swap_btn: $e("ramp_swap"),
        output: $e("ramp_output"),
    },
};

const midi = new MIDI("Gemini");
const gemini = new Gemini(midi);
const settings = new GemSettings();
let gemini_firmware_version = null;
let gemini_serial_number = null;
let gemini_hardware_revision = null;
let adc_calibration_backup = null;
let afe_calibration_backup = null;
let ramp_calibration_backup = null;

/*
  Factory calibrations are stored on GCS. This allows retrieving it.
*/

async function fetch_calibration(serial_no, type) {
    let response = await fetch(
        `https://storage.googleapis.com/files.winterbloom.com/calibrations/gemini/${serial_no}.${type}.json`
    );

    if (!response.ok) {
        console.warn(
            `Could not find ${type} calibration data for CPU ID ${serial_no}`
        );
        return null;
    }

    return await response.json();
}

async function check_for_backups() {
    adc_calibration_backup = await fetch_calibration(
        gemini_serial_number,
        "adc"
    );
    afe_calibration_backup = await fetch_calibration(
        gemini_serial_number,
        "afe"
    );
    ramp_calibration_backup = await fetch_calibration(
        gemini_serial_number,
        "ramp"
    );

    /* Check and see if there's a backup for  data. */
    if (!adc_calibration_backup && !afe_calibration_backup) {
        ui.restore_adc_calibration_btn.disabled = true;
        ui.restore_adc_calibration_btn.classList.remove("is-warning");
        ui.restore_adc_calibration_btn.classList.add("is-dark");
    }

    if (!ramp_calibration_backup) {
        ui.ramp.swap_btn.disabled = true;
        ui.ramp.swap_btn.classList.remove("is-warning");
        ui.ramp.swap_btn.classList.add("is-dark");
    }
}

async function restore_backup_calibration() {
    if (adc_calibration_backup) {
        settings.adc_gain_corr = Math.round(
            adc_calibration_backup.gain_error * 2048
        );
        settings.adc_offset_corr = Math.round(
            adc_calibration_backup.offset_error
        );
    }
    if (afe_calibration_backup) {
        settings.cv_gain_error = afe_calibration_backup.gain_error;
        settings.cv_offset_error = afe_calibration_backup.offset_error;
    }

    ui.settings_form.update();
}

async function check_for_new_firmware() {
    let gh = new GitHub();
    let release_info = null;
    try {
        release_info = await gh.get_latest_release(
            "wntrblm",
            "Castor_and_Pollux"
        );
    } catch (e) {
        console.log("Error while fetching latest firmware: ", e);
        return false;
    }

    if (gemini_firmware_version.includes(release_info.tag_name)) {
        return false;
    }

    return true;
}

$on(ui.connect_btn, "click", async function () {
    const connect_info_p = ui.connect_info.querySelector("p");

    ui.connect_btn.innerText = "Connecting";
    ui.connect_btn.disabled = true;
    connect_info_p.innerText = "";
    ui.connect_info.hidden = true;

    try {
        await midi.connect();
    } catch (err) {
        console.log(err);
        connect_info_p.innerText =
            "Couldn't connect, check connection and power and try again?";
        ui.connect_info.hidden = false;
        return;
    } finally {
        ui.connect_btn.innerText = "Connect";
        ui.connect_btn.disabled = false;
    }

    gemini_firmware_version = await gemini.get_version();
    ui.firmware_version.value = `${gemini_firmware_version}`;

    const serial_num_result = await gemini.get_serial_number();
    gemini_serial_number = serial_num_result.serial;
    gemini_hardware_revision = serial_num_result.revision;
    ui.serial_number.value = `${gemini_serial_number}`;
    ui.hardware_revision.value = `v${gemini_hardware_revision}`;

    ui.info_section.hidden = false;

    if (await check_for_new_firmware()) {
        console.log("Firmware too old, bailing. :(");
        ui.firmware_incompatible.hidden = false;
        return;
    }

    /* Load settings & update the form. */
    let loaded_settings = false;
    /* WebMIDI can sometimes inexplicably mess up SysEx messages, so try this a few times. */
    for (let attempt = 0; attempt < 10; attempt += 1) {
        try {
            Object.assign(settings, await gemini.load_settings());
            loaded_settings = true;
            console.log("Loaded settings", settings);
            break;
        } catch (err) {
            console.log("Retrying loading settings, got: ", err);
        }
    }

    if (!loaded_settings) {
        connect_info_p.innerText =
            "Couldn't load settings, check connection, power, and try resetting the module.";
        return;
    }

    ui.settings_form.update();
    draw_lfo_waveform();

    ui.connect_btn.disabled = true;
    ui.connect_btn.innerText = "Connected";
    ui.connect_info.hidden = true;
    ui.settings_section.hidden = false;
    ui.danger_section.hidden = false;

    check_for_backups();
});

$on(ui.save_btn, "click", async function () {
    ui.save_btn.disabled = true;
    ui.save_btn.innerText = "Saving...";
    console.log("Saving settings", settings);

    await gemini.save_settings(settings);
    gemini.soft_reset();

    ui.save_btn.classList.remove("is-primary");
    ui.save_btn.classList.add("is-success");
    ui.save_btn.innerText = "Saved!";

    await new Promise((resolve) => setTimeout(resolve, 2000));

    ui.save_btn.classList.add("is-primary");
    ui.save_btn.classList.remove("is-success");
    ui.save_btn.innerText = "Save";
    ui.save_btn.disabled = false;
});

$on(ui.restore_adc_calibration_btn, "click", async function () {
    await restore_backup_calibration();
});

/*
    Enable/disable dangerous settings.
*/
$on(ui.allow_danger, "change", function () {
    ui.dangerous_section_content.hidden = !ui.allow_danger.checked;
});

/*
  Form data binding and display logic
*/
ui.settings_form = new forms.Form(ui.settings_form_elem, settings);

/*
    Helper to draw the LFO waveform.
*/
function draw_lfo_waveform() {
    const get_waveform = (selection) => {
        selection = parseInt(selection);
        switch (selection) {
            // Triangle
            case 0:
                return (phase, frequency) =>
                    -1.0 + Math.abs(-2.0 + ((phase * frequency) % 1.0) * 4);
            // Sine
            case 1:
                return (phase, frequency) =>
                    Math.sin(2 * Math.PI * phase * frequency);
            // Sawtooth
            case 2:
                return (phase, frequency) =>
                    -1.0 + ((phase * frequency) % 1.0) * 2;
            // Square
            case 3:
                return (phase, frequency) =>
                    (phase * frequency) % 1.0 < 0.5 ? -1 : 1;
        }
    };

    const freq = 3;
    const lfo_1_f = get_waveform(settings.lfo_1_waveshape);
    const lfo_1_a = settings.lfo_1_factor;
    const lfo_2_f = get_waveform(settings.lfo_2_waveshape);
    const lfo_2_a = settings.lfo_2_factor;
    const lfo_2_r = settings.lfo_2_frequency_ratio;

    const lfo_func = (phase) => {
        return (
            lfo_1_f(phase, freq) * lfo_1_a +
            lfo_2_f(phase, freq * lfo_2_r) * lfo_2_a
        );
    };

    const canvas = ui.lfo_waveform_canvas;
    const ctx = canvas.getContext("2d");
    const padding = canvas.height * 0.2;

    ctx.clearRect(0, 0, canvas.width, canvas.height);

    ctx.strokeStyle = "rgb(94, 64, 158)";
    ctx.lineWidth = 10;
    ctx.lineCap = "round";
    ctx.lineJoin = "round";

    ctx.beginPath();
    for (let i = 0; i < 1.1; i += 1 / canvas.width) {
        let val = lfo_func(0.25 + i);
        val = (val + 1) / 2;
        ctx.lineTo(
            i * canvas.width,
            padding + val * (canvas.height - padding * 2)
        );
    }
    ctx.stroke();
}

draw_lfo_waveform();
for (let input of ui.settings_form.elem.querySelectorAll("[name*=lfo]")) {
    input.addEventListener("input", () => {
        window.requestAnimationFrame(draw_lfo_waveform);
    });
}

/*
    Tuning calibration helpers
*/
const tuning_info = {
    one_volt: 0,
    three_volt: 0,
    gain_error: 0,
    offset_error: 0,
};

$on(ui.tuning.measure_one, "click", async function () {
    tuning_info.one_volt = await gemini.read_adc_average(7, 20);
    ui.tuning.measure_one_result.value = tuning_info.one_volt;

    tuning_info.gain_error = null;
    tuning_info.offset_error = null;
    tuning_info.three_volt = null;

    ui.tuning.gain_error_result.value = tuning_info.gain_error;
    ui.tuning.offset_error_result.value = tuning_info.offset_error;
    ui.tuning.measure_three_result.value = tuning_info.three_volt;
    ui.tuning.save.disabled = true;
});

$on(ui.tuning.measure_three, "click", async function () {
    tuning_info.three_volt = await gemini.read_adc_average(7, 20);
    ui.tuning.measure_three_result.value = tuning_info.three_volt;

    if (!tuning_info.one_volt) return;

    const expected_low = gemini.volts_to_code(gemini_hardware_revision, 1.0);
    const expected_high = gemini.volts_to_code(gemini_hardware_revision, 3.0);
    console.log(expected_low, expected_high);
    const measured_low = tuning_info.one_volt;
    const measured_high = tuning_info.three_volt;
    tuning_info.gain_error = (
        (expected_high - expected_low) /
        (measured_high - measured_low)
    ).toFixed(4);
    tuning_info.offset_error = (
        measured_low * tuning_info.gain_error -
        expected_low
    ).toFixed(1);

    ui.tuning.gain_error_result.value = tuning_info.gain_error;
    ui.tuning.offset_error_result.value = tuning_info.offset_error;
    ui.tuning.save.disabled = false;
});

$on(ui.tuning.save, "click", function () {
    if (tuning_info.gain_error === null) return;
    if (tuning_info.offset_error === null) return;

    settings.cv_gain_error = tuning_info.gain_error;
    settings.cv_offset_error = tuning_info.offset_error;

    ui.settings_form.update();
    ui.save_btn.scrollIntoView();
    ui.save_btn.click();
});

/*
    Monitoring
*/
if (window.location.hash == "#monitor") {
    ui.monitoring.section.hidden = false;
}
$on(ui.monitoring.enable, "click", function () {
    gemini.enable_monitoring((msg) => {
        ui.monitoring.output.innerText = JSON.stringify(msg, undefined, 2);
    });
});

/*
    Ramp calibration swap
*/
if (window.location.hash == "#ramp") {
    ui.ramp.section.hidden = false;
}
$on(ui.ramp.swap_btn, "click", async function () {
    console.log(ramp_calibration_backup);

    const periods = Object.keys(ramp_calibration_backup["castor"])
        .sort((a, b) => parseInt(a, 10) - parseInt(b, 10))
        .reverse();

    for (let i = 0; i < periods.length; i++) {
        const period = periods[i];
        const period_number = parseInt(period, 10);
        let castor_value = ramp_calibration_backup["castor"][period];
        let pollux_value = ramp_calibration_backup["pollux"][period];

        // swap em
        [castor_value, pollux_value] = [pollux_value, castor_value];

        await gemini.write_lut_entry(
            i,
            period_number,
            castor_value,
            pollux_value
        );

        ui.ramp.output.innerText += `Set ${i}: period=${period}, castor=${pollux_value}, pollux=${castor_value}\n`;
    }

    await gemini.save_lut_table();

    ui.ramp.output.innerText += `Saved to NVM\n`;
});
